dll/win32/shell32/folders/CRecycleBin.cpp

```diff
@@ -171,16 +123,94 @@ BOOL WINAPI CBSearchRecycleBin(IN PVOID Context, IN HANDLE hDeletedFile)
     return ret;
 }
 
-static PIDLRecycleStruct * _ILGetRecycleStruct(LPCITEMIDLIST pidl)
+/*
+ * Recycle Bin folder
+ */
+
+HRESULT CRecyclerExtractIcon_CreateInstance(
+    LPCITEMIDLIST pidl, REFIID riid, LPVOID * ppvOut)
 {
-    LPPIDLDATA pdata = _ILGetDataPointer(pidl);
+    PIDLRecycleStruct *pFileDetails = _ILGetRecycleStruct(pidl);
+    if (pFileDetails == NULL)
+        goto error;
 
-    if (pdata && pdata->type == 0x00)
-        return (PIDLRecycleStruct*) & (pdata->u.crecycle);
+    // Try to obtain the file
+    SEARCH_CONTEXT Context;
+    Context.pFileDetails = pFileDetails;
+    Context.bFound = FALSE;
 
-    return NULL;
+    EnumerateRecycleBinW(NULL, CBSearchRecycleBin, (PVOID)&Context);
+    if (Context.bFound)
+    {
+        // This should happen any time
+        IRecycleBinFile* pRecycleFile = (IRecycleBinFile*)Context.hDeletedFile;
+
+        // Query the interface from the private interface
+        HRESULT hr = pRecycleFile->QueryInterface(riid, ppvOut);
+
+        // Close the file handle as we don't need it anymore
+        CloseRecycleBinHandle(Context.hDeletedFile);
+
+        return hr;
+    }
+    else
```

Comment:
Can you remove the `else` here please (it's not needed since the `if` block returns)? That'll make the `goto` much more elegant.

Reply:
yeah, i can do that

Reply:
In repect of the forward declaration, it's needed because if you keept in on that place a warning appears, that says that a function is used before it's declaration, i could add the declaration up, but this ways is more cleaner as any other use outside that file uses that function

Reply:
Done